src/relax/transform/dataflow_inplace.cc

```diff
@@ -0,0 +1,1017 @@
+  if (auto* var_node = target.as<VarNode>()) {
+    auto current_var = GetRef<Var>(var_node);
+    // if the var is live past this point, we can't use it for in-place computations anyway
+    if (live_ranges.count(current_var)) {
+      auto live_range = live_ranges.at(current_var);
+      if (live_range.second > binding_idx) {
+        return false;
+      }
+    }
+
+    // no entry for the current var -> it must be something external and we have to assume the worst
+    if (!alias_sets.count(current_var)) {
+      return false;
+    }
+    auto alias_set = alias_sets.at(current_var);
+    // -1 -> an external value and we must assume the worst
+    if (alias_set.count(-1)) {
+      return false;
+    }
+    std::vector<std::unordered_set<int>> sets_to_check = {alias_set};
+    std::unordered_set<int> indices_checked;
+    // If a possible alias is a tuple, we will also check for aliases of the members
+    // (possibly recursively)
+    for (int alias_idx : alias_set) {
+      if (!GatherSetsToCheckForLiveness(alias_sets, tuple_map, &sets_to_check, alias_idx)) {
+        return false;
+      }
+    }
+
+    for (Var other_var : currently_live) {
+      if (other_var.same_as(target)) {
+        continue;
+      }
+      // not represented = spooky unknown value that should be modeled by -1
+      if (!alias_sets.count(other_var) || !live_ranges.count(other_var)) {
+        continue;
+      }
+      // var is not live past this point => don't need to worry
+      if (live_ranges.at(other_var).second <= binding_idx) {
+        continue;
+      }
+      auto other_alias_set = alias_sets.at(other_var);
+      for (int alias_idx : other_alias_set) {
+        for (auto check_set : sets_to_check) {
+          if (check_set.count(alias_idx)) {
+            return false;
+          }
+        }
+      }
+    }
+    return true;
+  } else if (auto* tup_node = target.as<TupleNode>()) {
+    for (auto field : tup_node->fields) {
+      if (!InplaceConditionsMet(live_ranges, alias_sets, tuple_map, currently_live, field,
+                                binding_idx)) {
+        return false;
+      }
+    }
+    return true;
+  } else {
+    return true;
+  }
+}
+
+// this is obviously not a complete list
+static std::unordered_set<std::string> SUPPORTED_OPS = {"relax.add",      "relax.subtract",
+                                                        "relax.multiply", "relax.divide",
+                                                        "relax.nn.silu",  "relax.nn.relu"};
+bool OpSupportsInplace(const Op& op) { return SUPPORTED_OPS.count(op->name); }
+
+// Check for in-place eligibility:
+//  1. see if there's an arg big enough to hold the result
+//  2. see if the arg is live past the call
+//  3. see if the arg has an alias that's live past the call
+// If the conditions are met, record the index of that binding.
+// Returns two lists of lists:
+// 1. A list of bindings where at least one argument meets the in-place conditions and the *size*
```

Comment:
It would probably be neater to have a data structure to encode that info, yes. I didn't use one to make things easier for the FFI but that's not a very good reason. I'll use this suggestion.

Reply:
Yeah, good point on the FFI.  I think it's probably worth it for the readability, but the amount of bookkeeping necessary to expose an object through the FFI is more than I'd like.  (And duplicated between the C++ and Python sides.)  At some point, I'd like to take a look at how pybind11 exposes the C++ bindings and see if something similar could be adapted for TVM, as I like the way it can expose C++ types without requiring bindings on the python side.